sams-scala-tutorial/serialization-exceptions-and-memory-leaks-no-ws.md

```diff
@@ -13,14 +13,147 @@ Talk:
+    rdd.map { x => unSerializableClassHandler.func_1 }.count()
+
+    val unSerializableObjectHandler = NonSerilizableObject
+    rdd.map { x => unSerializableObjectHandler.func_1(x) }.count()
+
+    /* 2 above examples do not work, error information shown below:
+     * Exception in thread "main" org.apache.spark.SparkException: Task not serializable
+     * 	at org.apache.spark.util.ClosureCleaner$.ensureSerializable
+     * 
+     * The reason of the exception: all functions in RDD closure need to go through a process called ClosureCleaner.
+     * If ClosureCleaner is not able to cut off functions in the closure from context, it will try to serialize context.
+     * If the context cannot be serialized, this exception will happen.
+     * In both these 2 examples, closures have a reference outside respectively(unSerializableClassHandler and unSerializableObjectHandler).
+     * Obviously, because they do not extend Serializable, neither of them can be serialized. 
+     * */
+
+    /*How to solve this kind of exceptions? */
+
+    /*
+     * solution 1: Use reference outside and make object or class extend interface Serializable
+     * In this way, outer reference can be serialized by colusure cleaner and the task can be shipped from driver to data nodes.
+     * */
+    val SerializableClassHandler = new SerializableClass(1)
+    rdd.map { x => SerializableClassHandler.func_1 }.count()
+
+    val SerializableObjectHandler = SerilizableObject
+    rdd.map { x => SerializableObjectHandler.func_1(x) }.count()
+    /*
+     * Above 2 examples can be executed successfully because SerializableClassHandler and SerializableObjectHandler can be serialized by closure cleaner.
+     * */
+
+    /*
+     * Solution 2: Do not use outside reference, and instantiate a object for every element in RDD.
+     * In this way, for each element in RDD, a class will be instantiated or a local object will be referenced in its own node, so there is no worry about serialization.
+     * */
+    rdd.map { x =>
+      {
+        val handler = new NonSerializableClass(1)
+        handler.func_1
+      }
+    }.count()
+
+    rdd.map { x =>
+      {
+        val handler = NonSerilizableObject
+        handler.func_1(x)
+      }
+    }.count()
+    /*
+     * Above 2 examples works because no objects need to be serialized. 
+     * As I use nonSerializable object and class in RDD closure, whether serializable or not does not matter. 
+     * */
+
+    /*
+     * Solution 3: Do not use outside reference, and only use static functions.
+     * In scala, static functions only exist in Object so this method only work for Object
+     * */
+    rdd.map { x =>
+      {
+        NonSerilizableObject.func_1(x)
+      }
+    }.count()
+    /*
+     * The example works. Actually this method is the same as second example in solution 2.
+     * Method func_1 is directly from object without any outside reference.
+     * */
+
+    /*
```

Comment:
@QingyangKong  This tip is true, but I don't see any relevance to TSNE.  Please remove.